fuel-vm/src/constraints/reg_key.rs

```diff
@@ -0,0 +1,312 @@
+) -> (SystemRegisters<'_>, ProgramRegisters<'_>) {
+    let [zero, one, of, pc, ssp, sp, fp, hp, err, ggas, cgas, bal, is, ret, retl, flag, rest @ ..] = registers;
+    let r = SystemRegisters {
+        zero: RegMut(zero),
+        one: RegMut(one),
+        of: RegMut(of),
+        pc: RegMut(pc),
+        ssp: RegMut(ssp),
+        sp: RegMut(sp),
+        fp: RegMut(fp),
+        hp: RegMut(hp),
+        err: RegMut(err),
+        ggas: RegMut(ggas),
+        cgas: RegMut(cgas),
+        bal: RegMut(bal),
+        is: RegMut(is),
+        ret: RegMut(ret),
+        retl: RegMut(retl),
+        flag: RegMut(flag),
+    };
+    (r, ProgramRegisters(rest))
+}
+
+pub(crate) fn copy_registers(
+    read_registers: &SystemRegistersRef<'_>,
+    write_registers: &ProgramRegistersRef<'_>,
+) -> [Word; VM_REGISTER_COUNT] {
+    let mut out = [0u64; VM_REGISTER_COUNT];
+    out[..VM_REGISTER_READ_COUNT].copy_from_slice(&<[Word; VM_REGISTER_READ_COUNT]>::from(read_registers));
+    out[VM_REGISTER_READ_COUNT..].copy_from_slice(write_registers.0);
+    out
+}
+
+impl<'a> From<&'a SystemRegisters<'_>> for SystemRegistersRef<'a> {
+    fn from(value: &'a SystemRegisters<'_>) -> Self {
+        Self {
+            zero: Reg(value.zero.0),
+            one: Reg(value.one.0),
+            of: Reg(value.of.0),
+            pc: Reg(value.pc.0),
+            ssp: Reg(value.ssp.0),
+            sp: Reg(value.sp.0),
+            fp: Reg(value.fp.0),
+            hp: Reg(value.hp.0),
+            err: Reg(value.err.0),
+            ggas: Reg(value.ggas.0),
+            cgas: Reg(value.cgas.0),
+            bal: Reg(value.bal.0),
+            is: Reg(value.is.0),
+            ret: Reg(value.ret.0),
+            retl: Reg(value.retl.0),
+            flag: Reg(value.flag.0),
+        }
+    }
+}
+
+impl<'a> From<SystemRegisters<'a>> for SystemRegistersRef<'a> {
+    fn from(value: SystemRegisters<'a>) -> Self {
+        Self {
+            zero: Reg(value.zero.0),
+            one: Reg(value.one.0),
+            of: Reg(value.of.0),
+            pc: Reg(value.pc.0),
+            ssp: Reg(value.ssp.0),
+            sp: Reg(value.sp.0),
+            fp: Reg(value.fp.0),
+            hp: Reg(value.hp.0),
+            err: Reg(value.err.0),
+            ggas: Reg(value.ggas.0),
+            cgas: Reg(value.cgas.0),
+            bal: Reg(value.bal.0),
+            is: Reg(value.is.0),
+            ret: Reg(value.ret.0),
+            retl: Reg(value.retl.0),
+            flag: Reg(value.flag.0),
+        }
+    }
+}
+
+impl<'a> From<&'a ProgramRegisters<'_>> for ProgramRegistersRef<'a> {
+    fn from(value: &'a ProgramRegisters<'_>) -> Self {
+        Self(value.0)
+    }
+}
+
+impl<'a> From<ProgramRegisters<'a>> for ProgramRegistersRef<'a> {
+    fn from(value: ProgramRegisters<'a>) -> Self {
+        Self(value.0)
+    }
+}
+
+impl<'a> From<&SystemRegistersRef<'a>> for [Word; VM_REGISTER_READ_COUNT] {
+    fn from(value: &SystemRegistersRef<'a>) -> Self {
+        [
```

Comment:
Could you use `let SystemRegistersRef { and your variables here } = value;` please? It will help not to miss new registers=)